bdk-cli basics 00

logosstone opened the pull request:
This is a basics bdk-cli tutorial for quick reference.

logosstone committed:
Add bdk-cli introduction tutorial

rajarshimaitra commented:
Hey @waterst0ne is this still WIP?? Sorry took me some time to get back to the docs.. But really excited on this one.. Let me know if this is ready for review and will go through it..

logosstone commented:
> Hey @waterst0ne is this still WIP?? Sorry took me some time to get back to the docs.. But really excited on this one.. Let me know if this is ready for review and will go through it..

Hi @rajarshimaitra this is ready for review. I have a similar "2-of-3  Multisig" tutorial like this just about finished but I would just have to make some final changes. I'll  upload soon, thank you for reaching out.

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+---
+title: "Command Line intro to Bitcoin Wallet Dev"
```
A more verbose title, something like
```suggestion
title: "Command Line introduction to Bitcoin Wallet Development using bdk-cli"
```

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+- This short tutorial will expose you to the [`bdk library` ](https://docs.rs/bdk/latest/bdk/) and the practical knowledge needed for Bitcoin wallet development. As a consequence you will deepen your technical understanding about Bitcoin and the bdk library.
+
+- [BDK](https://github.com/bitcoindevkit/bdk-ffi) also has language bindigs for **Kotlin/Java, Swift, Python** which enable the use of BDK's **Rust** library as an API. You can later use these similar steps to create your own Bitcoin mobile, desktop or even WebApp by using the bdk-ffi language bindings. 
```
Lets mark the "language bindings" words with hyperlink instead of "BDK"

logosstone replied:
Good call.

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+# Bicoin Dev Kit - Command Line Intro to Bitcoin Wallet Development 
+
+- My goal for this tutorial is to introduce you to the powerful command-line program named `bdk-cli` a  useuful tool for Bitcoin wallet testing. You will be exposed to many of the basic techincal skills that go into creating and managing Bitcoin wallets. 
```
How about a hyperlink for bdk-cli pointing to the github repo?

logosstone replied:
yes.

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+# Bicoin Dev Kit - Command Line Intro to Bitcoin Wallet Development 
+
+- My goal for this tutorial is to introduce you to the powerful command-line program named `bdk-cli` a  useuful tool for Bitcoin wallet testing. You will be exposed to many of the basic techincal skills that go into creating and managing Bitcoin wallets. 
+- If you've read most of the "Mastering Bitcoin" book, this tutorial could serve as a stepping stone into your Bitcoin wallet development journey.
```
Same hyperlink love for "Mastering Bitcoin" would be nice too..

logosstone replied:
Yes 👍

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+### Outline of Tutorial and Installation notes:
+    
+### Brief Outline of Tutorial
+-  Step 1: Creating a mnemonic word list + XPRV (Extended Private Key)
```
Its mostly called as "Xprv"..

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+### `bdk-cli` installation:
+-  Download the [`bdk-cli` github repository locally](https://github.com/bitcoindevkit/bdk-cli.git)
+    -  Enter the folder `cd bdk-cli`
+    -  Install `cargo install --path . --features electrum,repl,compiler `
```
We don't really need `repl` and `compiler` for this tutorial..

rajarshimaitra replied:
We don't need repl and compiler..

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
````diff
@@ -0,0 +1,294 @@
+
+:+1: The output below confirms the command was successful.
+```
+bdk-cli 0.5.0
````
We are currently at 0.6.0..

rajarshimaitra replied:
better to update at v0.6.. Interface changed a bit..

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+
+ ---
+ 
```
A preview of `bdk-cli help` and a bit of intro on each command would be helpful here..

logosstone replied:
That makes sense in case the user wants to tinker . 💯

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+
+***
+## Step 5: Check the balance
```
A note here saying reader need to wait for a 1 block confirmation to see the balance update would be helpful..

logosstone replied:
ACK

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+:arrow_down: To create a PSBT (partially-signed-bitcoin-transaction) run  the command :arrow_down:
+
+:arrow_forward: `bdk-cli wallet --wallet wallet_name --descriptor $my_descriptor create_tx --to mkHS9ne12qx9pS9VojpwU5xtRd4T7X7ZUt:50000`
```
I am confused.. What is this address and why does it look like this??

logosstone replied:
I got this address from a test-net faucet and I'm not sure what kind of address is this. It shouldn't interfere with the tutorial in my opinion.

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+
+
+
+ - Verify transcation in the memory pool on testnet [Mempool-testnet!](https://mempool.space/testnet)
```
Finishing touch, run sync one more time and see your balance decreased..

logosstone replied:
Tip was added!

rajarshimaitra commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+- [bip-44: Multi-Account Hierarchy for Deterministic Wallets](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki)
+- [bip-84: Derivation scheme for P2WPKH based accounts](https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki)
+- [What are Descriptors and miniscript?](https://blog.summerofbitcoin.org/miniscript-policy-descriptors-hidden-powers-of-bitcoin/)
+- [Master Private  Key and Extended Private Key](https://bitcoin.stackexchange.com/questions/97242/bip39-tool-bip32-extended-private-key-vs-bip32-root-key)
```
One more link for PSBT would be nice..

logosstone replied:
Added it to the list. 
Than you  for all the recommendations, the tutorial looks much better now.

rajarshimaitra reviewed (changes requested):
ACK 35e029fa6105a6a04e27ae5f6187adee594dd2b3

Thanks for making this.. Really looking forward for your multisig tutorial.. If you are interested in writing some other cool ones with more complex scripts ping me in discord.. I have few ideas.. 

This one looks good, but I think it fits better in the tutorial section than in the blog posts.. 

Really like your video gif idea to show the commands in action.. This can be useful for more complex and involved tutorials.. 

For now some following nits..

logosstone commented on docs/_blog/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,294 @@
+permalink: "/_blog/bdk_cli_basics_00"
+---
+
+# Bicoin Dev Kit - Command Line Intro to Bitcoin Wallet Development 
```
Thank you.

rajarshimaitra commented:
Hey @waterst0ne is this ready for a second look? LMK, the new commit seems to be re adding the whole thing, and not a change set.. I tried manually checking few review comments but they don't seem to be update.. Having a change set of updates only would help in review a lot..

logosstone commented:
> 

Hi Raj, yes this is ready for a second review please. Regarding the commit not showing a change set updates I'll take note, thank you for mentioning it. Feel free to share more suggestions or corrections.

rajarshimaitra commented on docs/tutorials/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,348 @@
+
+## Tutorial Goals
+
+- My goal for this tutorial is to introduce you to [bdk-cli](https://github.com/bitcoindevkit/bdk-cli), a powerful command-line program used for Bitcoin wallet testing. You will be exposed to many of the basic techincal skills that go into creating and managing Bitcoin wallets. 
```
Nit
```suggestion
- The goal for this tutorial is to introduce you to [bdk-cli](https://github.com/bitcoindevkit/bdk-cli), a powerful command-line program used for Bitcoin wallet testing. Made using the more generic [BDK](https://docs.rs/bdk/latest/bdk/) wallet library.  You will be exposed to many of the basic technical skills that go into creating and managing Bitcoin wallets. 
```

rajarshimaitra commented on docs/tutorials/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,348 @@
+:arrow_forward: `bdk-cli wallet --wallet wallet_name  --descriptor $my_descriptor get_balance `
+
+:::tip
+Note: The balance will only show after the transaction has been confirmed in a block at least once.
```
Now I am thinking if this is correct. We should also be able to see unconfirmed balances right? Maybe we can also show how the balance transformed from confirmed to unconfirmed??

rajarshimaitra commented on docs/tutorials/bdk_cli_basics_00.md:
````diff
@@ -0,0 +1,348 @@
+:+1: The output below confirms the command was successful:
+```
+{
+  "satoshi": 100000
````
With version 6, this balance structure updates.. And we now have a lot more details..

rajarshimaitra commented on docs/tutorials/bdk_cli_basics_00.md:
```diff
@@ -0,0 +1,348 @@
+
+:arrow_down: To create a PSBT (partially-signed-bitcoin-transaction) run  the command :arrow_down:
+
+:arrow_forward: `bdk-cli wallet --wallet wallet_name --descriptor $my_descriptor create_tx --to mkHS9ne12qx9pS9VojpwU5xtRd4T7X7ZUt:50000`
```
Hmm I think its better to work with old formats than to burn test coins.. I am just thinking if this could confuse readers.. But I guess they could learn something interesting too if they notice it looks weird..

rajarshimaitra reviewed (changes requested):
ACK a9c0b21c9999cbc940308cd0afe2a124493e222c. Thanks for the updates.. 

You also need to include the new tutorial path here to get it displayed in the website..
https://github.com/bitcoindevkit/bitcoindevkit.org/blob/689236703f372b1ef82795000a7027f35248b883/docs/.vuepress/config.js#L49-L59 

Took me some time to figure, but you forgot to delete the file from `_blog` folder. That's why previous changes aren't outdated yet.. Can fix that on next push.. 

few more nits, and update requests..

thunderbiscuit commented:
@waterst0ne are you still working on this? I'm looking to clear up the PR backlog for the website this week. Let me know what you want to do with this.

If it's almost there I can read/fix a bit and merge, otherwise if it's too far back in terms of API we'll just close it.

logosstone commented:
> @waterst0ne are you still working on this? I'm looking to clear up the PR backlog for the website this week. Let me know what you want to do with this.
> 
> If it's almost there I can read/fix a bit and merge, otherwise if it's too far back in terms of API we'll just close it.

I added some minor changes and also added windows powershell commands.
I think it's ready now.

thunderbiscuit commented:
I cleaned up a lot of the structure without changing the content and squashed the commits.

thunderbiscuit commented:
@rajarshimaitra do you mind approving this? The PR can't be merged because you had requested changes but I think it's ready to go now.

rajarshimaitra reviewed (approved):
Sorry took me some time.. 

ACK 7677a3b6ad392dda44047d1594395a9d46f5fc23
